wled00/FX.h

```diff
@@ -363,6 +363,7 @@ typedef struct Segment {
     };
     uint8_t startY;  // start Y coodrinate 2D (top); there should be no more than 255 rows
     uint8_t stopY;   // stop Y coordinate 2D (bottom); there should be no more than 255 rows
+    uint16_t transitionprogress; // current transition progress 0 - 0xFFFF
```

Comment:
This could be a `static` variable as it will be modified in each `handleTransition()` call at the start of `service()` loop.

Reply:
what is the current behaviour if for example the palette of one segment is changed with 10s transition, and after 5s palette of a second segment is changed? Will that stop transition of the first segment? i.e. are transition start/stop per segment or global?

Reply:
The transition time is bound to `strip` ATM but each segment may start transition at its own (point in) time. So each segment should transition on its own (that was my goal when implementing current transitions, compared to previous, limited transitions) independent from others.

Reply:
> This could be a `static` variable as it will be modified in each `handleTransition()` call at the start of `service()` loop.

Rather than make it static, I'd say put it into `strip`. 

Static member attributes of a class are always a good source of confusion when trying to read code written by someone else - because a 'static' member is technicially not even part of the object you work with. It survives `delete segment`, and changing it in one object instances also changes the value in all other instances.

Reply:
> Rather than make it static, I'd say put it into `strip`.

That would be a worse choice IMO. It can be considered in the same way as _vLength, etc in speed improvements branch by @DedeHai . It is used as a speedup, pre-calculated value.
If you think this will cause confusion, keep it as an instance member rather than `strip` member.
But that is just my opinion, no need to take it into account.

Reply:
If individual segment transition is to be kept/extended in the future, this should not be static but per segment. Also there is no RAM impact currently as there are 3 bytes of padding added in that space anyway.

Reply:
> this should not be static but per segment

Look at the `Segment::beginDraw()` in another PR (#4138). Uses the same principle for `virtualLength()` etc.
If this PR is embedded there, there is no need for different logic regarding calculation of `progress` variable. Still, if effect function runs for more than 1ms then the result may not be accurate (within that function) regardless of where this substitution/pre-calculation comes from. `beginDraw()` is called at the start of effect drawing and `progress()` is used within `setPixelColor()` when blending is enabled (and more so in #4158 ).

Reply:
That is good point. I guess the question is whether progress should be per frame, which was my intention, or per frame per segment, which is what you are suggesting if I understand that right. 
Doing it per frame would keep segment progress aligned i.e. first and last segment rendered use the same progress. I am not sure what the best solution is with regards to sync (or frame based super sync) if ever implemented as I do not understand it yet in detail.

Reply:
`progress()` determines the state where transition is. It should not be "frame" based. It only depends on time since transition (for that segment) started. BTW it is possible to have multiple segments in transition with differing `progress()` value.

Reply:
what is the reason `progress()` should depend on the order of segment processing or when does this give an advantage aver calculating it once per frame? If for example there are lots of segments with heavy FX calculation (>1ms) later segments would already be further in the transition in the same frame, resulting in different palette or brightness for example. That does not appear to be the correct approach IMHO. It would only be right if startTransition() is called with the same time delay, which I don't think it is. Using the 'per segment' update is more consistent code though, as you said. I think your suggestion of following the `Segment::beginDraw()` is a good compromise.

Reply:
> what is the reason `progress()` should depend on the order of segment processing

It should not be *dependant on the order of segment processing*. Each "segment" (if you want it) has its own "next time" for update in the form of effect function of return value. So some effects may run at independent frame rate. Look at Android and a few others (i.e. Game of life).

What matters is the time difference between the start of transition and current time (in regards to the transition duration). This is not frame or segment based. It is true that segment holds both start time and transition duration and these two are unique to each segment. This was the prime reason to have `progress()` as a function that calculates how far transition has progressed. If you impose a limit that progress is calculated at the start of frame (for optimisation for speed) then it needs to be pre-calculated for each segment separately. Hence the suggestion for `beginDraw()`.

Why do you think that having different progress value for is wrong? Each segment operates independently and displays effect and/or palette independently of other segments. Even if overlaid. It is totally acceptable for two segments to be at different progress level.

And it is quite easy to achieve that.

Reply:
There are different scenarios, your reasoning seems sound to me. My scenario was this: have 10 segments, all displaying the same FX, change palette on all of them at the same time -> a transition starts on all segments. Palette change may now not be simulatanuous on all segments, the last ones may already have progressed further. Is my thinking here correct for current code? I am not arguing against your suggestion to keep it 'per segment', which I think is the way to go, just trying to understand where a 'once per frame' would be better and where it would be worse.

Reply:
It is correct and in most cases this will be the case (same start for all segments). But not in all cases (and not all segments may have same effect and/or palette).

So, if you have an exception, treat it as a regular behaviour (not all segments may start transition at the same time).